Platform/CoffeelakeBoardPkg/Library/Stage2BoardInitLib/Stage2BoardInitLib.c

```diff
@@ -2128,7 +2106,11 @@ UpdateCpuNvs (
   ///
   /// Get CPU Init Data Hob
   ///
-  GuidHob = GetNextGuidHob (&gCpuInitDataHobGuid, LdrGlobal->FspHobList);
+  GuidHob = NULL;
+  FspHobList = GetFspHobListPtr ();
+  if (FspHobList != NULL) {
+    GuidHob = GetNextGuidHob (&gCpuInitDataHobGuid, FspHobList);
+  }

```

Comment:
How about update the GetNextGuidHob() to return NULL when the second parameter is NULL.
so we could still use GetNextGuidHob() in this case.

Reply:
I thought about it, but this is to keep EDKII interface as it is.

  **If Guid is NULL, then ASSERT().
  If HobStart is NULL, then ASSERT().**

```
In MdePkg\Include\Library\HobLib.h

/**
  Returns the next instance of the matched GUID HOB from the starting HOB.

  This function searches the first instance of a HOB from the starting HOB pointer.
  Such HOB should satisfy two conditions:
  its HOB type is EFI_HOB_TYPE_GUID_EXTENSION and its GUID Name equals to the input Guid.
  If there does not exist such HOB from the starting HOB pointer, it will return NULL.
  Caller is required to apply GET_GUID_HOB_DATA () and GET_GUID_HOB_DATA_SIZE ()
  to extract the data section and its size info respectively.
  In contrast with macro GET_NEXT_HOB(), this function does not skip the starting HOB pointer
  unconditionally: it returns HobStart back if HobStart itself meets the requirement;
  caller is required to use GET_NEXT_HOB() if it wishes to skip current HobStart.

  If Guid is NULL, then ASSERT().
  If HobStart is NULL, then ASSERT().

  @param  Guid          The GUID to match with in the HOB list.
  @param  HobStart      A pointer to a Guid.

  @return The next instance of the matched GUID HOB from the starting HOB.

**/
VOID *
EFIAPI
GetNextGuidHob (
  IN CONST EFI_GUID         *Guid,
  IN CONST VOID             *HobStart
  );
```

Reply:
Yes, EDKII will ASSERT when the second parameter is NULL since it is not expected.
Comparing with new code, did you see issue with below code?  The FspHob should not be NULL else it should be bug.
GuidHob = GetNextGuidHob (&gCpuInitDataHobGuid, GetFspHobListPtr ());

Reply:
GetFspHobListPtr () is NULL until MemoryInit is done. Caller MUST guarantee that GetNextGuidHob() parameters are not NULL.

Reply:
MemoryInit must be done in stage2 code.  Prefer to use previous logic to make code simple since this patch mainly changes how to get FSP HOB pointer (from global data or from a function), not change FSP HOB itself.
But I am ok if you insist to have a check every time use FSP HOB pointer.

Reply:
MemoryInit is done in Stage1B. Refer to Stage1B.c line#442. Until the line, FspHobList is NULL.
This patch added a safe code to make sure NOT NULL. If we change GetNextGuidHob(), caller codes would be simpler but we need to change MdePkg\Include\Library\HobLib.h interface description/definition.

Reply:
It could not be NULL in this code Stage2BoardInitLib.c, why we newly add the check for FspHobList. That's my concern.

Reply:
Right, it shouldn't be NULL in Stage2 until ClearFspHob(). This is to add a safe code since we have a chance to have NULL FspHobList.

Reply:
From my viewpoint, it is not necessary to add such check: a) Simplify the caller. b) Capture such logic error early by assertion (If it asserts, the code logic much be fixed why we get FSP hob while FspHobList is NULL.)

Reply:
Your first comment was to make GetNextGuidHob() return NULL when the second parameter is NULL instead of ASSERTION. Now, it's back to ASSERTION.

Reply:
At that time I had thought GetNextGuidHob() would hang if NULL is passed to it when I saw the new check was added.

Reply:
Let's this big patch firstly since adding additional check doesn't harm anything. we could change it back later if required.